include/lsst/afw/image/PhotoCalib.h

```diff
@@ -0,0 +1,426 @@
+#include "lsst/afw/math/ChebyshevBoundedField.h"
+#include "lsst/afw/geom/Point.h"
+#include "lsst/afw/geom/Box.h"
+#include "lsst/afw/table/Source.h"
+#include "lsst/afw/table/io/Persistable.h"
+
+namespace lsst { namespace afw { namespace image {
+
+/// A value and its error.
+struct Quantity {
+    Quantity(double value, double sigma): value(value), sigma(sigma) {}
+    double const value;
+    double const sigma;
+};
+
+class PhotoCalib : public table::io::PersistableFacade<PhotoCalib>, public table::io::Persistable {
+public:
+    // no move or copy
+    PhotoCalib(PhotoCalib const &) = delete;
+    PhotoCalib(PhotoCalib &&) = delete;
+    PhotoCalib & operator=(PhotoCalib const &) = delete;
+    PhotoCalib & operator=(PhotoCalib &&) = delete;
+
+    /**
+     * Create a empty, zeroed calibration.
+     */
+    PhotoCalib(): PhotoCalib(0) {}
+
+    /**
+     * Create a non-spatially-varying calibration.
+     *
+     * @param[in]  countsMag0       The constant counts/magnitude zero point (counts at magnitude 0).
+     * @param[in]  countsMag0Sigma  The error on the zero point.
+     * @param[in]  bbox             The bounding box on which this PhotoCalib is valid.
+     */
+    explicit PhotoCalib(double countsMag0, double countsMag0Sigma=0,
+                        afw::geom::Box2I const & bbox=afw::geom::Box2I()):
+        _countsMag0(countsMag0), _countsMag0Sigma(countsMag0Sigma), _isConstant(true) {
+            ndarray::Array<double, 2, 2> coeffs = ndarray::allocate(ndarray::makeVector(1, 1));
+            coeffs[0][0] = countsMag0;
+            _zeroPoint = std::make_shared<afw::math::ChebyshevBoundedField>(
+                                    afw::math::ChebyshevBoundedField(bbox, coeffs));
+        }
+
+    /**
+     * Create a spatially-varying calibration.
+     *
+     * @param[in]  zeroPoint      The spatially varying photometric zero point.
+     * @param[in]  countsMag0Sigma  The error on the zero point.
+     */
+    PhotoCalib(std::shared_ptr<afw::math::BoundedField> zeroPoint, double countsMag0Sigma=0):
+        _zeroPoint(zeroPoint), _countsMag0(computeCountsMag0(zeroPoint)),
+        _countsMag0Sigma(countsMag0Sigma), _isConstant(false) {}
+
+    /**
+     * Create a calibration with a pre-computed mean. Primarily for de-persistence.
+     *
+     * @param[in]  countsMag0      The mean of the zeroPoint() over its bounding box.
+     * @param[in]  zeroPoint       The spatially varying photometric zero point.
+     * @param[in]  countsMag0Sigma The error on the zero point.
+     * @param[in]  isConstant      Is this PhotoCalib spatially constant?
+     */
+    PhotoCalib(double countsMag0, double countsMag0Sigma,
+               std::shared_ptr<afw::math::BoundedField> zeroPoint,
+               bool isConstant):
+        _zeroPoint(zeroPoint), _countsMag0(countsMag0), _countsMag0Sigma(countsMag0Sigma),
+        _isConstant(isConstant) {}
+
+    //@{
+    /**
+     * Convert counts in ADU to maggies.
```

Comment:
I already deleted my notes on the subject because I thought you weren't interested, but going from memory:

The danger of having lots of convenience methods is that it not only makes more work for future class maintainers, it makes the API harder to learn by the user, who needs to carefully compare each method to figure out which one they need (case in point, it's impossible to have a constructive discussion about this without a list of the overloads in hand). A good rule of thumb is to only add a convenience overload if you expect *lots* of people to find it easier than doing without it, and to leave it out if only a handful will get a strong advantage.

I think the output parameter used in number 7 is more a holdover from C than something that's still justified in C++ (I assume the original motivation was that arrays need a lot more manual management than container objects do). BTW, what's the behavior if `result` doesn't have the right dimensions for the output? The documentation doesn't say.

I assume versions 6 and 7 are equivalent to just iterating over the catalog and calling 5 on each record? Then, per the rule of thumb above, they probably don't add enough extra functionality to be worth the human overhead. (Yes, this is the exact opposite of what you said above re: 5 being redundant with 4; consider it an alternative point of view.)

A similar argument would apply to version 8 -- would it be more useful to have an update-inplace version of 5, and let the user decide how many records they want to update?